source/validate_cfg.cpp

```diff
@@ -424,14 +424,63 @@ spv_result_t PerformCfgChecks(ValidationState_t& _) {
       for (auto edge : postdom_edges) {
         edge.first->SetImmediatePostDominator(edge.second);
       }
-      /// calculate back edges.
+      /// Calculate back edges.
       DepthFirstTraversal(
           function.pseudo_entry_block(),
           function
               .AugmentedCFGSuccessorsFunctionIncludingHeaderToContinueEdge(),
           ignore_block, ignore_block, [&](cbb_ptr from, cbb_ptr to) {
             back_edges.emplace_back(from->id(), to->id());
           });
+
+      // Perform a DepthFirstTraversal that ensures CFG nesting depth is valid.
+      const int cfg_depth_limit = 1023;
+      bool cfg_depth_limit_violated = false;
+      int depth = 0;
+      std::unordered_map<cbb_ptr, int> block_depth;
+      DepthFirstTraversal(
+          function.first_block(), function.AugmentedCFGSuccessorsFunction(),
+          [&](cbb_ptr b) {
+            // Header block of a loop.
+            if (b->is_type(kBlockTypeLoop)) {
+              // at this point we can set the depth for the merge block and
+              // continue block of the loop.
+              const auto& construct = function.FindConstructForEntryBlock(b);
+              const auto& merge_block = construct.exit_block();
+              // since it's a loop, the corresponding construct is just the
+              // continue construct: A vector with 1 element.
+              const auto& continue_construct =
+                  construct.corresponding_constructs()[0];
+              const auto& continue_block = continue_construct->entry_block();
+              block_depth[b] = depth;
+              block_depth[merge_block] = depth;
+              block_depth[continue_block] = depth + 1;
+              ++depth;
+            } else if (b->is_type(kBlockTypeHeader)) {
+              // Header block of a selection.
+              const auto& construct = function.FindConstructForEntryBlock(b);
+              const auto& merge_block = construct.exit_block();
+              block_depth[b] = depth;
+              block_depth[merge_block] = depth;
+              ++depth;
+            } else if (block_depth.find(b) == block_depth.end()) {
+              block_depth[b] = depth;
+            }
```

Comment:
I'm thinking that fixing this bug might give us the correct solution. But I'll try to think of counter examples.